File: cli/src/core/TelemetryBuffer.test.ts
import { statSync } from "node:fs";
import { access, mkdir, mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { getJolliMemoryDir } from "../Logger.js";
import {
	appendTelemetryEvent,
	clearTelemetryBuffer,
	MAX_BYTES,
	MAX_EVENTS,
	readTelemetryEvents,
	replaceTelemetryEvents,
	type TelemetryEnvelope,
} from "./TelemetryBuffer.js";

let cwd: string;

const env = (over: Partial<TelemetryEnvelope> = {}): TelemetryEnvelope => ({
	schemaVersion: 1,
	eventName: "recall_performed",
	surface: "cli",
	surfaceVersion: "1.2.0",
	installId: "install-abc",
	sessionId: "sess-1",
	os: "darwin",
	arch: "arm64",
	runtimeVersion: "node-22.5.0",
	env: "prod",
	tsIso: "2026-06-20T00:00:00.000Z",
	accountId: null,
	properties: {},
	...over,
});

const queueFile = (dir: string) => join(getJolliMemoryDir(dir), "telemetry-queue.ndjson");

beforeEach(async () => {
	cwd = await mkdtemp(join(tmpdir(), "telemetry-buffer-"));
});
afterEach(async () => {
	await rm(cwd, { recursive: true, force: true });
});

describe("TelemetryBuffer", () => {
	it("round-trips appended events (missing file starts empty)", async () => {
		expect(await readTelemetryEvents(cwd)).toEqual([]);
		appendTelemetryEvent(cwd, env({ properties: { hit: true } }));
		appendTelemetryEvent(cwd, env({ eventName: "search_performed" }));
		const events = await readTelemetryEvents(cwd);
		expect(events).toHaveLength(2);
		expect(events[0].properties).toEqual({ hit: true });
		expect(events[1].eventName).toBe("search_performed");
	});

	it("creates the jollimemory dir on first append", async () => {
		appendTelemetryEvent(cwd, env());
		await expect(access(queueFile(cwd))).resolves.toBeUndefined();
	});

	it("compacts in place when the file exceeds MAX_BYTES (append stays bounded)", async () => {
		// Append well past MAX_BYTES so the append-time guard must compact;
		// without it the file would grow unbounded when the flusher never runs.
		const big = env({ properties: { pad: "x".repeat(200) } });
		const perEvent = `${JSON.stringify(big)}\n`.length;
		const count = Math.ceil((MAX_BYTES * 2) / perEvent);
		for (let i = 0; i < count; i++) appendTelemetryEvent(cwd, big);
		expect(statSync(queueFile(cwd)).size).toBeLessThanOrEqual(MAX_BYTES);
		expect((await readTelemetryEvents(cwd)).length).toBeLessThanOrEqual(MAX_EVENTS);
	});

	it("read caps to the newest MAX_EVENTS (drop-oldest)", async () => {
		for (let i = 0; i < MAX_EVENTS + 10; i++) {
			appendTelemetryEvent(cwd, env({ installId: `i-${i}` }));
		}
		const events = await readTelemetryEvents(cwd);
		expect(events).toHaveLength(MAX_EVENTS);
		// oldest 10 dropped → first kept is i-10, last is i-509
		expect(events[0].installId).toBe("i-10");
		expect(events[MAX_EVENTS - 1].installId).toBe(`i-${MAX_EVENTS + 9}`);
	});

	it("skips a corrupt line but keeps the rest", async () => {
		const dir = getJolliMemoryDir(cwd);
		await mkdir(dir, { recursive: true });
		await writeFile(
			queueFile(cwd),
			`${JSON.stringify(env({ installId: "good-1" }))}\n{torn json\n\n${JSON.stringify(env({ installId: "good-2" }))}\n`,
			"utf-8",
		);
		const events = await readTelemetryEvents(cwd);
		expect(events.map((e) => e.installId)).toEqual(["good-1", "good-2"]);
	});

	it("replace overwrites and caps; empty array removes the file", async () => {
		appendTelemetryEvent(cwd, env({ installId: "old" }));
		await replaceTelemetryEvents(cwd, [env({ installId: "kept-1" }), env({ installId: "kept-2" })]);
		expect((await readTelemetryEvents(cwd)).map((e) => e.installId)).toEqual(["kept-1", "kept-2"]);

		await replaceTelemetryEvents(cwd, []);
		expect(await readTelemetryEvents(cwd)).toEqual([]);
		await expect(access(queueFile(cwd))).rejects.toThrow();
	});

	it("replace caps an oversized input to MAX_EVENTS", async () => {
		const big = Array.from({ length: MAX_EVENTS + 5 }, (_, i) => env({ installId: `b-${i}` }));
		await replaceTelemetryEvents(cwd, big);
		const events = await readTelemetryEvents(cwd);
		expect(events).toHaveLength(MAX_EVENTS);
		expect(events[0].installId).toBe("b-5");
	});

	it("clear removes the buffer (idempotent when already absent)", async () => {
		appendTelemetryEvent(cwd, env());
		await clearTelemetryBuffer(cwd);
		expect(await readTelemetryEvents(cwd)).toEqual([]);
		await expect(clearTelemetryBuffer(cwd)).resolves.toBeUndefined();
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/core/telemetry/TelemetryBufferTest.kt
package ai.jolli.jollimemory.core.telemetry

import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import java.io.File
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir

class TelemetryBufferTest {
    @TempDir
    lateinit var tempDir: File

    private val cwd: String get() = tempDir.absolutePath

    private fun env(
        installId: String = "install-1",
        eventName: String = "recall_performed",
        properties: Map<String, Any?> = mapOf("hit" to true),
    ) = TelemetryEnvelope(
        schemaVersion = 1,
        eventId = "22222222-2222-4222-8222-222222222222",
        eventName = eventName,
        surface = "intellij",
        surfaceVersion = "1.0.0",
        installId = installId,
        sessionId = "s1",
        os = "mac",
        arch = "arm64",
        runtimeVersion = "jvm-21",
        env = "local",
        tsIso = "2026-06-20T00:00:00.000Z",
        accountId = null,
        properties = properties,
    )

    private fun queueFile() = File("$cwd/.jolli/jollimemory/telemetry-queue.ndjson")

    @Test
    fun `append then read round-trips (missing file starts empty)`() {
        TelemetryBuffer.readLines(cwd) shouldHaveSize 0
        TelemetryBuffer.append(cwd, env(eventName = "app_installed"))
        TelemetryBuffer.append(cwd, env(eventName = "search_performed"))
        TelemetryBuffer.readLines(cwd) shouldHaveSize 2
        val parsed = TelemetryBuffer.read(cwd)
        parsed shouldHaveSize 2
        parsed[0].eventName shouldBe "app_installed"
        parsed[1].eventName shouldBe "search_performed"
    }

    @Test
    fun `compacts in place when the file exceeds MAX_BYTES`() {
        val big = env(properties = mapOf("pad" to "x".repeat(500)))
        val perEvent = com.google.gson.Gson().toJson(big).length + 1
        val count = ((TelemetryBuffer.MAX_BYTES * 2) / perEvent).toInt() + 1
        repeat(count) { TelemetryBuffer.append(cwd, big) }
        (queueFile().length() <= TelemetryBuffer.MAX_BYTES) shouldBe true
        TelemetryBuffer.readLines(cwd) shouldHaveSize TelemetryBuffer.MAX_EVENTS
    }

    @Test
    fun `read caps to the newest MAX_EVENTS`() {
        for (i in 0 until TelemetryBuffer.MAX_EVENTS + 10) {
            TelemetryBuffer.append(cwd, env(installId = "i-$i"))
        }
        val lines = TelemetryBuffer.readLines(cwd)
        lines shouldHaveSize TelemetryBuffer.MAX_EVENTS
        // oldest 10 dropped → first kept is i-10
        (lines.first().contains("\"installId\":\"i-10\"")) shouldBe true
    }

    @Test
    fun `read skips a corrupt line but keeps the rest`() {
        val f = queueFile()
        f.parentFile.mkdirs()
        f.writeText(
            "${com.google.gson.Gson().toJson(env(installId = "good-1"))}\n" +
                "{torn json\n\n" +
                "${com.google.gson.Gson().toJson(env(installId = "good-2"))}\n",
            Charsets.UTF_8,
        )
        TelemetryBuffer.read(cwd).map { it.installId } shouldBe listOf("good-1", "good-2")
    }

    @Test
    fun `replaceLines overwrites and caps, and empty input removes the file`() {
        TelemetryBuffer.append(cwd, env(installId = "old"))
        TelemetryBuffer.replaceLines(cwd, listOf("""{"a":1}""", """{"b":2}"""))
        TelemetryBuffer.readLines(cwd) shouldHaveSize 2

        TelemetryBuffer.replaceLines(cwd, emptyList())
        TelemetryBuffer.readLines(cwd) shouldHaveSize 0
        queueFile().exists() shouldBe false
    }

    @Test
    fun `clear removes the buffer and is idempotent`() {
        TelemetryBuffer.append(cwd, env())
        TelemetryBuffer.clear(cwd)
        TelemetryBuffer.readLines(cwd) shouldHaveSize 0
        TelemetryBuffer.clear(cwd) // no throw when already absent
    }
}
